### PR TITLE
feat(components): Update InputDate to customize date format, calendar icon and displayed value

### PR DESCRIPTION
**Pull request by nshakoori**
## Motivations
To make the `InputDate` component more flexible to implement the new Schedule design

## Changes
Added `AtlantisContext` for web components.
Added the `emptyValuePlaceholder` prop to `InputDate` to have the option of displaying a string instead of a date.
Added the `showIcon` prop to `InputDate` to make showing the calendar icon optional
Updated `DatePicker` to apply `dateFormat` from `AtlantisContext` if it's available

### Added

The `InputDate` component now has the following two features:
- hide/show the calendar icon
- display a value other then a date

### Changed

No changes in existing functionality 
[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

File: packages/components/src/AtlantisContext/AtlantisContext.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  timeZone: Intl.DateTimeFormat().resolvedOptions().timeZone,
+  isOnline: true,
+  onLogError: _ => {
+    return;
+  },
+  floatSeparators: { group: ",", decimal: "." },
+  currencySymbol: "$",
+  headerHeight: 0,
+  locale: "en",
+  setHeaderHeight: _ => {
```
Comment (MichaelParadis) [inline packages/components/src/AtlantisContext/AtlantisContext.tsx:79]: Anything relating to headerheight isn't needed. It is very specific to the header bar we have in React Native

File: packages/components/src/InputDate/InputDate.tsx [hunk 2]
```diff
@@ -57,13 +62,14 @@ export function InputDate(inputProps: InputDateProps) {
...
 
-        const suffix = {
-          icon: "calendar",
-          ...(onClick && {
-            onClick: onClick,
-            ariaLabel: "Show calendar",
-          }),
-        } as Suffix;
+        const suffix =
+          inputProps.showIcon === undefined || inputProps.showIcon
```
Comment (MichaelParadis) [inline packages/components/src/InputDate/InputDate.tsx:66]: I think this can be re-written as `inputProps !=== false` and on the doc string for show icon add

`@default true`

File: packages/components/src/AtlantisContext/AtlantisContext.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+   * Time zone used in converting the date and time formats
+   */
+  readonly timeZone: string;
+
+  /**
+   * Determines how the component behaves when the app is online or offline.
+   * This is set to `true` by default. If your usage doesn't require a different
+   * experience, you can leave this as-is.
+   */
+  readonly isOnline: boolean;
```
Comment (MichaelParadis) [inline packages/components/src/AtlantisContext/AtlantisContext.tsx:25]: React native specific so this can be removed. `onLogError` also isn't needed right now

File: packages/components/src/AtlantisContext/AtlantisContext.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+   * The currency symbol Atlantis components will use
+   */
+  readonly currencySymbol: string;
+
+  /**
+   * The `headerHeight` property represents the height of the app header in Atlantis.
+   * It plays a crucial role in determining the positioning of various elements within the app.
+   * By accurately defining this value, Atlantis can effectively calculate the layout and alignment of its components.
+   */
+  readonly headerHeight: number;
```
Comment (MichaelParadis) [inline packages/components/src/AtlantisContext/AtlantisContext.tsx:35]: Header height should be removed because it is only used for stuff related to React Navigation in react native

Comment (nshakoori) [reply packages/components/src/AtlantisContext/AtlantisContext.tsx]: Done! Missed this one.

File: packages/components/package.json [hunk 5]
```diff
@@ -73,11 +73,13 @@
...
+    "date-fns": "^2.0.0",
+    "date-fns-tz": "^2.0.0"
   },
   "browserslist": [
     "last 1 version",
     "> 1%",
     "IE 10"
   ]
-}
+}
```
Comment (MichaelParadis) [inline packages/components/package.json:85]: I think saving this with eslint or a formatter enabled will undo this change.

File: packages/components/src/AtlantisContext/AtlantisContext.test.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  atlantisContextDefaultValues,
+  useAtlantisContext,
+} from "./AtlantisContext";
+
+const providerValues: AtlantisContextProps = {
+  dateFormat: "MM/DD/YYYY",
+  timeFormat: "hh:mm a",
+  timeZone: "America/Edmonton",
+  locale: "en",
+  isOnline: false,
```
Comment (MichaelParadis) [inline packages/components/src/AtlantisContext/AtlantisContext.test.tsx:16]: Some of these are no longer needed

File: packages/components/src/FormField/FormField.tsx [hunk 7]
```diff
@@ -44,11 +44,12 @@ export function FormField(props: FormFieldProps) {
     onValidation,
     onKeyUp,
     clearable = "never",
+    emptyValuePlaceholder,
   } = props;
 
   const {
     control,
-    formState: { errors },
+    formState: { errors, isDirty },
```
Comment (MichaelParadis) [inline packages/components/src/FormField/FormField.tsx:52]: I think this change and the change in the below use effect is no longer needed

File: packages/components/src/AtlantisContext/AtlantisContext.test.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  onLogError: _ => {
+    return;
+  },
+  floatSeparators: { decimal: ".", group: "," },
+  currencySymbol: "€",
+  headerHeight: 50,
+  setHeaderHeight: _ => {
+    return;
+  },
```
Comment (MichaelParadis) [inline packages/components/src/AtlantisContext/AtlantisContext.test.tsx:25]:
```suggestion
```

File: packages/components/src/AtlantisContext/AtlantisContext.test.tsx [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+const providerValues: AtlantisContextProps = {
+  dateFormat: "MM/DD/YYYY",
+  timeFormat: "hh:mm a",
+  timeZone: "America/Edmonton",
+  locale: "en",
+  isOnline: false,
+  onLogError: _ => {
+    return;
+  },
```
Comment (MichaelParadis) [inline packages/components/src/AtlantisContext/AtlantisContext.test.tsx:19]:
```suggestion
```

File: packages/components/src/FormField/FormFieldTypes.ts [hunk 10]
```diff
@@ -226,4 +226,6 @@ export interface FormFieldProps extends CommonFormFieldProps {
    * highlights the the field red if an error message shows up.
    */
   readonly validations?: RegisterOptions;
+
+  readonly emptyValuePlaceholder?: string;
```
Comment (MichaelParadis) [inline packages/components/src/FormField/FormFieldTypes.ts:230]:
```suggestion
```

File: packages/components/src/FormField/FormField.tsx [hunk 11]
```diff
@@ -68,7 +69,9 @@ export function FormField(props: FormFieldProps) {
   );
 
   useEffect(() => {
-    if (value != undefined) {
+    if (emptyValuePlaceholder && !isDirty) {
+      setValue(controlledName, emptyValuePlaceholder);
+    } else if (value != undefined) {
```
Comment (MichaelParadis) [inline packages/components/src/FormField/FormField.tsx]:
```suggestion
    if (value != undefined) {
```

File: packages/components/src/FormField/FormField.tsx (hunk 7 above)
Comment (MichaelParadis) [inline packages/components/src/FormField/FormField.tsx]:
```suggestion
    formState: { errors },
```

File: packages/components/src/FormField/FormField.tsx [hunk 12]
```diff
@@ -44,11 +44,12 @@ export function FormField(props: FormFieldProps) {
     onValidation,
     onKeyUp,
     clearable = "never",
+    emptyValuePlaceholder,
```
Comment (MichaelParadis) [inline packages/components/src/FormField/FormField.tsx]:
```suggestion
```

File: packages/components/src/InputDate/InputDate.tsx [hunk 13]
```diff
@@ -56,14 +67,17 @@ export function InputDate(inputProps: InputDateProps) {
...
+          !value,
+        );
+        const suffix =
+          inputProps.showIcon !== false
+            ? ({
+                icon: "calendar",
+                ariaLabel: "Show calendar",
+                onClick: onClick && onClick,
+              } as Suffix)
+            : {};
```
Comment (MichaelParadis) [inline packages/components/src/InputDate/InputDate.tsx:80]:
```suggestion
            : undefined;
```

Minor cleanup

File: packages/components-native/package.json [hunk 14]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@jobber/components-native",
-  "version": "0.60.0",
```
Comment (MichaelParadis) [inline packages/components-native/package.json:3]: Not sure where these changes to components-native came from. I think we should up this change and the one on the change log. I assume a conflict was resolved incorrectly or something

File: package-lock.json [hunk 15]
```diff
@@ -45756,13 +45756,15 @@
...
+        "date-fns": "^2.0.0",
+        "date-fns-tz": "^2.0.0",
         "react": "^18",
         "react-dom": "^18"
       }
     },
     "packages/components-native": {
       "name": "@jobber/components-native",
-      "version": "0.60.0",
+      "version": "0.61.0",
```
Comment (MichaelParadis) [inline package-lock.json:45767]: This should probably be reverted. Lerna does a lot of magic so it may have caused the changes to components native but we should undo them to reduce the chance something gets broken that we don't expect

Comment (MichaelParadis) [inline package-lock.json:45767]:
```suggestion
      "version": "0.60.0",
```

File: packages/components/src/InputDate/InputDate.tsx [hunk 16]
```diff
@@ -38,6 +38,17 @@ interface InputDateProps
...
+  /**
+   * Whether to show the calendar icon
+   * @default true
+   */
+  readonly showIcon?: boolean;
+
+  /**
+   * Text to display instead of a date value
+   */
+  readonly emptyValuePlaceholder?: string;
```
Comment (MichaelParadis) [inline packages/components/src/InputDate/InputDate.tsx:51]: To match the mobile version I think this prop should be updated to be
`emptyValueLabel` (sorry for missing this before).

File: packages/components/src/InputDate/InputDate.test.tsx [hunk 17]
```diff
@@ -196,6 +199,145 @@ describe("when InputDate is used within a Modal", () => {
...
+  const date = "11/11/2011";
+  const changeHandler = jest.fn();
+  const { queryByDisplayValue } = render(
+    <InputDate value={new Date(date)} onChange={changeHandler} />,
+  );
+
+  expect(queryByDisplayValue(date)).toBeInTheDocument();
+});
+
+it("should display emptyValuePlaceholder when set", () => {
```
Comment (MichaelParadis) [inline packages/components/src/InputDate/InputDate.test.tsx:328]: test names to be updated to `emptyValueLabel`

File: packages/components/src/InputDate/InputDate.test.tsx [hunk 18]
```diff
@@ -196,6 +199,145 @@ describe("when InputDate is used within a Modal", () => {
...
+    );
+    const input = getByDisplayValue(date);
+
+    fireEvent.keyUp(input, { key: "ArrowDown" });
+
+    expect(getByText("15")).toBeInTheDocument();
+  });
+});
+
+it("should display the selected date when emptyValuePlaceholder is undefined", () => {
```
Comment (MichaelParadis) [inline packages/components/src/InputDate/InputDate.test.tsx]: Same test name update here

**Review by MichaelParadis — APPROVED** [review]
LGTM! Great job on this!
